### PR TITLE
Refactored the concept of ElasticSearchRefresh code to class CatalogTraversal.

### PR DESCRIPTION
**Pull request by ajoymajumdar**
Deprecated class ElasticSearchRefresh.

### CODE REVIEW

File: metacat-main/src/main/java/com/netflix/metacat/main/api/v1/TagController.java [hunk 1]
```diff
@@ -559,12 +555,10 @@ private void removeResourceTags(final TagRemoveRequestDto tagRemoveRequestDto) {
         switch (name.getType()) {
             case CATALOG:
                 //catalog service will throw exception if not found
-                final CatalogDto catalogDto = this.catalogService.get(name, GetCatalogServiceParameters.builder()
+                this.catalogService.get(name, GetCatalogServiceParameters.builder()
                     .includeDatabaseNames(false).includeUserMetadata(false).build());
-                if (catalogDto != null) {
-                    this.tagService.removeTags(name, tagRemoveRequestDto.getDeleteAll(),
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/api/v1/TagController.java]: The catalogDto will always NOT be null or the tag API can handle null internally?

File: metacat-main/src/main/java/com/netflix/metacat/main/configs/ElasticSearchConfig.java [hunk 2]
```diff
@@ -179,4 +180,40 @@ public ElasticSearchRefresh elasticSearchRefresh(
             registry
         );
     }
+
+    /**
+     * ElasticSearch refresh action.
+     *
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/configs/ElasticSearchConfig.java]: Think the refresh refers to the catalogTraversalAction, can we make the comment consistent?

File: metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+     * Does a sweep across all catalogs to refresh the same data in elastic search.
+     */
+    public void process() {
+        processCatalogs(getCatalogNames());
+    }
+
+    /**
+     * Does a sweep across given catalogs to refresh the same data in elastic search.
+     *
+     * @param catalogNames catalog anmes
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java]: typo names

File: metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+            .map(s -> QualifiedName.ofDatabase(catalogName, s)).collect(Collectors.toList());
+        _process(qNames, () -> _processDatabases(QualifiedName.ofCatalog(catalogName), qNames), "processDatabases",
+            true, 1000);
+    }
+
+    @SuppressWarnings("checkstyle:methodname")
+    private void _process(final List<QualifiedName> qNames, final Supplier<ListenableFuture<Void>> supplier,
+                          final String requestName, final boolean delete, final int queueSize) {
+        if (isTraversalAlreadyRunning.compareAndSet(false, true)) {
+            final long start = registry.clock().wallTime();
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java]: How long will this traversal run? If it's a very long running process, we should set the lock at the individual catalog level.

File: metacat-main/src/main/java/com/netflix/metacat/main/services/search/ElasticSearchCatalogTraversalAction.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+@Slf4j
+public class ElasticSearchCatalogTraversalAction implements CatalogTraversalAction {
+    private final Config config;
+    private final DatabaseService databaseService;
+    private final TableService tableService;
+    private final ElasticSearchUtil elasticSearchUtil;
+    private final UserMetadataService userMetadataService;
+    private final TagService tagService;
+    private final MetacatEventBus eventBus;
+    private Registry registry;
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/search/ElasticSearchCatalogTraversalAction.java]: final registry?

File: metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversalAction.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+     * @param tables list of tables
+     */
+    default void applyTables(CatalogTraversal.Context context, List<Optional<TableDto>> tables) { }
+
+    /**
+     * Called when the catalog traversal ends.
+     *
+     * @param context traversal context
+     */
+    default void done(CatalogTraversal.Context context) { }
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversalAction.java]: What else would we do in this done method? If just delete the unmarked entities, we should name the delete as an method to make it more explicit.

File: metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java (hunk 4 above)
Comment (ajoymajumdar) [reply metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java]: At present, I am not changing the behavior. We should revisit this later.

File: metacat-main/src/main/java/com/netflix/metacat/main/configs/ElasticSearchConfig.java (hunk 2 above)
Comment (ajoymajumdar) [reply metacat-main/src/main/java/com/netflix/metacat/main/configs/ElasticSearchConfig.java]: Will do.

**Pull request merged**